### ISSUE TITLE
How to fine-tune bi-encoder

### ISSUE THREAD

**Issue by carschno** [UKPLab/sentence-transformers#1078]
This is a documentation request rather than an issue.

I have the following scenario:
I have a gold data set of query-document pairs (~30k pairs, with ~2700 distinct queries) that have been annotated as relevant/irrelevant. On average, I have 2.2 relevant documents per query.
Language is English.

I am wondering about the best approach to fine-tune a pre-trained model towards that data.

My current approach is roughly following the [train_bi-encoder-v2.py example](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder-v2.py) for MSMARCO, but diverging in a few points. Here are the essential steps:

- convert the training/test data into triplets of the form `<query>, <relevant doc>, <irrelevant doc>`: `InputExample(texts=[query_text, pos_doc_text, neg_doc_text])`
- start with a pre-trained bi-encoder model, such as `paraphrase-MiniLM-L6-v2`: `model = SentenceTransfomer("sentence-transformers/paraphrase-MiniLM-L6-v2"`; I assume that is a better choice than a vanilla BERT model.
- fit the model using `MultipleNegativesRankingLoss`
- use the `InformationRetrievalEvaluator` for evaluating results

For evaluation with the `InformationRetrievalEvaluator`, I convert the test triplets so that `corpus` comprises all documents in the test triplet set, and the `relevant_docs` comprises the relevant docs per query (e.g. `{"q1": {"doc1", "doc2", ...}}`).

Accuracy@1 remains mostly below the random baseline (0.5), and hardly improves during training. accuracy@10, however, comes close to 0.9. Precision and recall, on the other hand are very low (~0.48 prec@1, 0.19 prec@10, 0.25 recall@1, 0.82 recall@10). 

I am wondering about a few things:
- Why does the example use `MultipleNegativesRankingLoss` instead of [TripletLoss](https://www.sbert.net/docs/package_reference/losses.html#tripletloss), whereas I understand the literature suggests the latter? When I replace it, all evaluation scores from the `InformationRetrievalEvaluator` are significantly worse 
- What is the best format for evaluation with my data set? Given the structure of my gold data set (<query>-<document> annotations), the data provides no information about most query-document combinations. Hence, the evaluation scores might not be very meaningful.

**Comment by nreimers** [comment]
Hi @carschno 

Accuracy@k / prec@k / recall@k are a bit different to understand than what you write: 
Given a query and a corpus of e.g. 10k elements: Accuracy@1 measures if you retrieve a relevant hit on the first position. So random baseline would be 1/10000. Acc@10 measures if you have a relevant hit in the first 10 results.

Precision@10: Measures how many relevant hits you have in the 10 hits returned. If you only have 2 relevant hits in total per query, Prec@10 cannot be higher than 2/10 and Prec@100 cannot be higher than 2/100.
So using Prec@k is not really meaningful.

The best is too look at MRR and / or MAP. These are the most relevant measures for IR.

MultipleNegativesRankingLoss usually works much much much better than TripletLoss. It is also the loss that is most commonly used to train state-of-the-art systems.
TripletLoss has the disadvantage that you just have 1 negative per query. If the triplets are too easy, the model stops learnings. MultipleNegativesRankingLoss  uses all other positives and negatives in a batch as additional negatives. So when you have a batch size of 64, then each query has 63 + 64 = 127 negatives. Learning is much stronger in this case.

Regarding evaluation:
This is an issue with IR evaluation. There are far too many combinations which cannot be judged. You can continue what you have and use MAP or MRR of the InformationRetrievalEvaluator. Once you have a model, you can use it to retrieve for your eval queries further documents which have not been annotated and annotate them.

But usually, when you compare similar methods (e.g. dense retrieval methods), the ranking of systems will not change if you add further annotations per query. The scores will improve, but the ranking will roughly stay the same (e.g. all systems improve 0.1 in MRR).
